libensemble/gen_classes/aposmm.py

```diff
@@ -0,0 +1,123 @@
+            gen_specs["out"] = [
+                ("x", float, n),
+                ("x_on_cube", float, n),
+                ("sim_id", int),
+                ("local_min", bool),
+                ("local_pt", bool),
+            ]
+            gen_specs["persis_in"] = ["x", "f", "local_pt", "sim_id", "sim_ended", "x_on_cube", "local_min"]
+        if not persis_info:
+            persis_info = add_unique_random_streams({}, 2, seed=4321)[1]
+        super().__init__(History, persis_info, gen_specs, libE_info, **kwargs)
+        if not self.persis_info.get("nworkers"):
+            self.persis_info["nworkers"] = gen_specs["user"]["max_active_runs"]  # ??????????
+        self.all_local_minima = []
+        self._ask_idx = 0
+        self._last_ask = None
+        self._tell_buf = None
+        self._n_buffd_results = 0
+        self._n_total_results = 0
+        self._told_initial_sample = False
+
+    def _slot_in_data(self, results):
+        """Slot in libE_calc_in and trial data into corresponding array fields. *Initial sample only!!*"""
+        self._tell_buf["f"][self._n_buffd_results] = results["f"]
+        self._tell_buf["x"][self._n_buffd_results] = results["x"]
+        self._tell_buf["sim_id"][self._n_buffd_results] = results["sim_id"]
+        self._tell_buf["x_on_cube"][self._n_buffd_results] = results["x_on_cube"]
+        self._tell_buf["local_pt"][self._n_buffd_results] = results["local_pt"]
+
+    @property
+    def _array_size(self):
+        """Output array size must match either initial sample or N points to evaluate in parallel."""
+        user = self.gen_specs["user"]
+        return user["initial_sample_size"] if not self._told_initial_sample else user["max_active_runs"]
+
+    @property
+    def _enough_initial_sample(self):
+        """We're typically happy with at least 90% of the initial sample, or we've already told the initial sample"""
+        return (
+            self._n_buffd_results >= self.gen_specs["user"]["initial_sample_size"] - 10
+        ) or self._told_initial_sample
+
+    def ask_numpy(self, num_points: int = 0) -> npt.NDArray:
+        """Request the next set of points to evaluate, as a NumPy array."""
+        if (self._last_ask is None) or (
+            self._ask_idx >= len(self._last_ask)
+        ):  # haven't been asked yet, or all previously enqueued points have been "asked"
+            self._ask_idx = 0
+            self._last_ask = super().ask_numpy(num_points)
+            if self._last_ask[
+                "local_min"
+            ].any():  # filter out local minima rows, but they're cached in self.all_local_minima
+                min_idxs = self._last_ask["local_min"]
+                self.all_local_minima.append(self._last_ask[min_idxs])
+                self._last_ask = self._last_ask[~min_idxs]
+        if num_points > 0:  # we've been asked for a selection of the last ask
+            results = np.copy(
+                self._last_ask[self._ask_idx : self._ask_idx + num_points]
```

Comment:
When we are in this situation (perhaps in error) where APOSMM does not give us enough points (e.g. the current ask is for 6, but APOSMM has only supplied one) then this gets filled up with empty points. This should be handled somehow. If it should never arise, then as a error condition.